synapse/lib/trigger.py

```diff
@@ -1,69 +1,262 @@
+import os
 import logging
+import contextlib
+import collections
+import contextvars
+import dataclasses
+
+from typing import Optional
+
+import synapse.exc as s_exc
+import synapse.common as s_common
+
+import synapse.lib.msgpack as s_msgpack
 
 logger = logging.getLogger(__name__)
 
+Conditions = set((
+    'tag:add',
+    'tag:del',
+    'node:add',
+    'node:del',
+    'prop:set',
+))
+
+RecursionDepth = contextvars.ContextVar('RecursionDepth', default=0)
+
 class Triggers:
 
-    def __init__(self):
-        self._trig_list = []
-        raise FIXME
+    TRIGGERS_DB_NAME = 'triggers'
+    RECURSION_LIMIT = 16
 
-    def clear(self):
-        '''
-        Clear all previously registered triggers
+    @dataclasses.dataclass
+    class Rule:
+        ver: int  # version: must be 0
+        cond: str  # condition from above list
+        user: str  # username
+        storm: str  # story query
+        form: Optional[str] = dataclasses.field(default=None) # form name
+        tag: Optional[str] = dataclasses.field(default=None) # tag name
+        prop: Optional[str] = dataclasses.field(default=None) # property name
+
+        def __post_init__(self):
```

Comment:
can we use `mesg='...'` syntax for the synerr's we are raising here?

Reply:
I don't understand.

Reply:
got it